### DISCUSSION TITLE
Implementation of line recognition in class TextMeasurer

### DISCUSSION THREAD

**Discussion by steffi3011** [SixLabors/Fonts#202] [Ideas] [answered]
Hello,
we use this library for a project in which we calculate complex layouts using the TextMeasurer class from SixLabors.Fonts. Among other things, we need the number of lines from the generated layout.
Unfortunately there is currently no way in the library to output this number. As far as I can see, this can be done using a value that is already mapped. In the GlyphLayout class already exists a corresponding value, which indicates whether the glyph is in a new line.
It could look like this:
int usedLines = glyphsToRender.Count(x => x.StartOfLine);
Since I think this is a useful feature in general, I wanted to ask if you could implement a feature with the function for line recognition. Alternatively, I would also agree to implement a corresponding feature and create a pull request.
Thanks in advance
Stefanie

**Comment by JimBobSquarePants** [comment, marked as answer]
The algorithm would be slightly different as you can have multiple overlapping glyphs at the start of a line (emoji) but we capture the grapheme index so can still do what you want. 

I'd almost think that rather than simply exposing a Count we could actually expose a type containing a collection of font rectangles each representing a line. 

@tocsoft What do you think? We do have all that information captured I believe?

**Reply by tocsoft** [reply]
hmmm the more of these cases I see for specialised helpers around text measurement or layout access makes me more and more inclined to just expose `TextLayout` and `GlyphLayout` directly and let people fill there boots with their own extension methods/helpers around them... we then don't need to maintain them move forwards and we can concentrate on the core competencies of the library which is font loading and text layout.

I would even be inclined to remove some of the methods on `TextMeasurer` we already have... like the `TryMeasureCharacterBounds` for example (if memory serves it was only added for a specific user) and instead just move the code over to the samples repo as and example on utilising the TextLayout api for custom interrogation.

**Reply by ANahr** [reply]
I guess for us both approaches would work.
We just want to know how many lines have been created. Please note that while you can do that heuritically by Glyph position there is no certain way to know how many lines have been layed out simply by observing positions.

**Comment by steffi3011** [comment]
Hello, has anything been done here? And if so, which of the two approaches has been implemented?

**Reply by JimBobSquarePants** [reply]
Not as yet. We did a major (complete) rewrite of the layout engine at the end of last year to support bidi text so this wasn't really looked at. Looking at the problem now I think I would prefer adding the following methods.

```c#
TextMeasurer.CountLines(string text, TextOptions options)
TextMeasurer.CountLines(ReadOnlySpan<char> text, TextOptions options)
```
My reasoning behind this is twofold. 
1. I'm very confident now in the accuracy of our results.
2. I'm still not comfortable exposing the internal `GlyphLayout` struct and `TextLayout` should definitely (IMO) stay internal while we finish implementing missing text shapers.

Would be really easy for someone to PR now.
